Demo application in Go with a server and CLI

mugdha-adhav opened the pull request:
## Summary
* Adding Makefile
* Adding gitignore file
* Adding go mod files
* CLI application using Cobra library
* Minimal web server using net/http packag net/http package
* Logging for Go application using uber/zap library

mugdha-adhav committed:
Demo application in Go with a server and CLI

* Adding Makefile
* Adding gitignore file
* Adding go mod files
* CLI application using Cobra library
* Minimal web server using net/http packag net/http package
* Logging for Go application using uber/zap library

mugdha-adhav committed:
Add Make target for building and running server separately

mugdha-adhav committed:
Update README and gitignore files

mbtamuli commented on go.mod:
```diff
@@ -0,0 +1,10 @@
+module PhilomathesInc/demo-ci-pipelines
```
```suggestion
module github.com/PhilomathesInc/demo-ci-pipelines
```
Import paths without dots in the first path element are reserved for the standard library and toolchain.

See - https://github.com/golang/go/issues/32819, also - https://golang.org/ref/mod

mbtamuli replied:
Although you might get away with using this for now, it is not standard, and might break with any future release of go, since this behaviour is not promised.

mugdha-adhav replied:
Yes, will fix it. Thanks for pointing out.

mbtamuli commented on server/server.go:
```diff
@@ -0,0 +1,58 @@
+	return func(w http.ResponseWriter, req *http.Request) {
+		c.Logger.Info("received request to check the server health")
+		w.Header().Add("Content-Type", "application/json")
+		response, err := json.Marshal(Healthy)
```
There is no need to Marshal the string here. As you're setting the `Content-Type` header to `application/json` already, the response being expected is simply JSON data and the contents of your string `Healthy ` is actually JSON data. You just need to convert it to a byte slice.

```
# Your server response, when you're sending the marshalled data
$ curl -s localhost:9091/health | jq .
"{\"status\":\"ok\"}"
$ curl -s localhost:9091/health | jq .status
jq: error (at <stdin>:0): Cannot index string with string "status"

# As a demonstration for the correct JSON response, see GitHub API
$ curl -s https://api.github.com/users/mugdha-adhav | jq .
{
  "login": "mugdha-adhav",
  "id": 37295843,
  "node_id": "MDQ6VXNlcjM3Mjk1ODQz",
  "avatar_url": "https://avatars.githubusercontent.com/u/37295843?v=4",
  "gravatar_id": "",
  "url": "https://api.github.com/users/mugdha-adhav",
  "html_url": "https://github.com/mugdha-adhav",
  "followers_url": "https://api.github.com/users/mugdha-adhav/followers",
  "following_url": "https://api.github.com/users/mugdha-adhav/following{/other_user}",
  "gists_url": "https://api.github.com/users/mugdha-adhav/gists{/gist_id}",
  "starred_url": "https://api.github.com/users/mugdha-adhav/starred{/owner}{/repo}",
  "subscriptions_url": "https://api.github.com/users/mugdha-adhav/subscriptions",
  "organizations_url": "https://api.github.com/users/mugdha-adhav/orgs",
  "repos_url": "https://api.github.com/users/mugdha-adhav/repos",
  "events_url": "https://api.github.com/users/mugdha-adhav/events{/privacy}",
  "received_events_url": "https://api.github.com/users/mugdha-adhav/received_events",
  "type": "User",
  "site_admin": false,
  "name": "Mugdha Adhav",
  "company": "@acquia ",
  "blog": "https://mugdha.dev",
  "location": "Remote",
  "email": null,
  "hireable": true,
  "bio": "Programmer & Gamer",
  "twitter_username": "AdhavMugdha",
  "public_repos": 12,
  "public_gists": 1,
  "followers": 8,
  "following": 26,
  "created_at": "2018-03-12T11:48:32Z",
  "updated_at": "2021-07-08T05:47:17Z"
}
$ curl -s https://api.github.com/users/mugdha-adhav | jq .login
"mugdha-adhav"

# If you don't marshall and simply send the byte slice of the string, it behaves similar to GitHub API
$ curl -s localhost:9091/health | jq .
{
  "status": "ok"
}
$ curl -s localhost:9091/health | jq .status
"ok"
```

mugdha-adhav replied:
Great catch @mbtamuli
Fixing it.
And thanks for the elaborate description, it helped understand and test things quickly.

mbtamuli commented on cli/cmd/health.go:
```diff
@@ -0,0 +1,65 @@
+		c.Logger.Error("could not read the response from the server", zap.String("error", err.Error()))
+		return
+	}
+	if err := json.Unmarshal(body, &serverResponse); err != nil {
```
You are `Unmarshal`ing to a `string` here. 

Based on my other comment about your response, don't send the `Marshal`ed. But since your response data iss valid JSON and you'd be setting the `Content-Type`, you will need to `Unmarshal` the received JSON. But it won't to `Unmarshal`ed to a `string`. Maybe something like

```go
type serverResponse struct {
	Status string `json:"status"`
}
// (...truncated...)
sR := &serverResponse{}
if err := json.Unmarshal(body, &sR); err != nil {
	c.Logger.Error("could not unmarshal the response from the server", zap.String("error", err.Error()))
	return
}
if sR.Status == "ok" {
	fmt.Println("The server responded with healthy status")
} else {
	fmt.Println("The server is unhealthy")
}
fmt.Println("Response from server: ", sR.Status)
```



To read - https://blog.golang.org/json

mugdha-adhav replied:
Fixed. Thanks for writing out the solution.

mugdha-adhav committed:
Changing module name to github.com/PhilomathesInc/demo-ci-pipelines

mugdha-adhav committed:
Remove JSON marshalling and unmarshalling server response

Write []byte to the ResponseWritter instead of json marshalling
Unmarshal server response in a struct in CLI

mbtamuli reviewed (approved):
LGTM 👍🏼
